sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_command_buffer.cpp

```diff
@@ -0,0 +1,750 @@
+
+  uint32_t SrcOriginX = ur_cast<uint32_t>(SrcOrigin.x);
+  uint32_t SrcOriginY = ur_cast<uint32_t>(SrcOrigin.y);
+  uint32_t SrcOriginZ = ur_cast<uint32_t>(SrcOrigin.z);
+
+  uint32_t SrcPitch = SrcRowPitch;
+  if (SrcPitch == 0)
+    SrcPitch = ur_cast<uint32_t>(Region.width);
+
+  if (SrcSlicePitch == 0)
+    SrcSlicePitch = ur_cast<uint32_t>(Region.height) * SrcPitch;
+
+  uint32_t DstOriginX = ur_cast<uint32_t>(DstOrigin.x);
+  uint32_t DstOriginY = ur_cast<uint32_t>(DstOrigin.y);
+  uint32_t DstOriginZ = ur_cast<uint32_t>(DstOrigin.z);
+
+  uint32_t DstPitch = DstRowPitch;
+  if (DstPitch == 0)
+    DstPitch = ur_cast<uint32_t>(Region.width);
+
+  if (DstSlicePitch == 0)
+    DstSlicePitch = ur_cast<uint32_t>(Region.height) * DstPitch;
+
+  uint32_t Width = ur_cast<uint32_t>(Region.width);
+  uint32_t Height = ur_cast<uint32_t>(Region.height);
+  uint32_t Depth = ur_cast<uint32_t>(Region.depth);
+
+  const ze_copy_region_t ZeSrcRegion = {SrcOriginX, SrcOriginY, SrcOriginZ,
+                                        Width,      Height,     Depth};
+  const ze_copy_region_t ZeDstRegion = {DstOriginX, DstOriginY, DstOriginZ,
+                                        Width,      Height,     Depth};
+
+  std::vector<ze_event_handle_t> ZeEventList;
+  UR_CALL(getEventsFromSyncPoints(CommandBuffer, NumSyncPointsInWaitList,
+                                  SyncPointWaitList, ZeEventList));
+
+  ur_event_handle_t LaunchEvent;
+  UR_CALL(EventCreate(CommandBuffer->Context, nullptr, true, &LaunchEvent));
+  LaunchEvent->CommandType = CommandType;
+
+  ZE2UR_CALL(zeCommandListAppendMemoryCopyRegion,
+             (CommandBuffer->ZeCommandList, Dst, &ZeDstRegion, DstPitch,
+              DstSlicePitch, Src, &ZeSrcRegion, SrcPitch, SrcSlicePitch,
+              LaunchEvent->ZeEvent, ZeEventList.size(), ZeEventList.data()));
+
+  urPrint("calling zeCommandListAppendMemoryCopyRegion() with"
+          "  ZeEvent %#lx\n",
+          ur_cast<std::uintptr_t>(LaunchEvent->ZeEvent));
+
+  // Get sync point and register the event with it.
+  *SyncPoint = CommandBuffer->GetNextSyncPoint();
+  CommandBuffer->RegisterSyncPoint(*SyncPoint, LaunchEvent);
+  return UR_RESULT_SUCCESS;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL
+urCommandBufferCreateExp(ur_context_handle_t Context, ur_device_handle_t Device,
+                         const ur_exp_command_buffer_desc_t *CommandBufferDesc,
+                         ur_exp_command_buffer_handle_t *CommandBuffer) {
+  // Force compute queue type for now. Copy engine types may be better suited
+  // for host to device copies.
+  uint32_t QueueGroupOrdinal =
+      Device->QueueGroup[ur_device_handle_t_::queue_group_info_t::type::Compute]
+          .ZeOrdinal;
+
+  ZeStruct<ze_command_list_desc_t> ZeCommandListDesc;
+  ZeCommandListDesc.commandQueueGroupOrdinal = QueueGroupOrdinal;
+
+  ze_command_list_handle_t ZeCommandList;
+  ZE2UR_CALL(zeCommandListCreate, (Context->ZeContext, Device->ZeDevice,
```

Comment:
for future optimization: have a pool of lists and reuse them.

Reply:
Good idea. I've added a comment above as a TODO for future, and created an internal ticket for our team to look at this